services/meta/data.go

```diff
@@ -369,16 +369,48 @@ func (data *Data) CreateShardGroup(database, policy string, timestamp time.Time)
 		return nil
 	}
 
+	startTime := timestamp.Truncate(rpi.ShardGroupDuration).UTC()
+	endTime := startTime.Add(rpi.ShardGroupDuration).UTC()
+	if endTime.After(time.Unix(0, models.MaxNanoTime)) {
+		// Shard group range is [start, end) so add one to the max time.
+		endTime = time.Unix(0, models.MaxNanoTime+1)
+	}
+
+	for i := range rpi.ShardGroups {
+		if rpi.ShardGroups[i].Deleted() {
+			continue
+		}
+		startI := rpi.ShardGroups[i].StartTime
+		endI := rpi.ShardGroups[i].EndTime
+		if rpi.ShardGroups[i].Truncated() {
+			endI = rpi.ShardGroups[i].TruncatedAt
+		}
+
+		// shard_i covers range [start_i, end_i)
+		// We want the largest range [startTime, endTime) such that all of the following hold:
+		//   startTime <= timestamp < endTime
+		//   for all i, not { start_i < endTime && startTime < end_i }
+		// Assume the above conditions are true for shards index < i, we want to modify startTime,endTime so they are true
+		// also for shard_i
+
+		// It must be the case that either endI <= timestamp || timestamp < startI, because otherwise:
+		// startI <= timestamp < endI means timestamp is contained in shard I
+		if !timestamp.Before(endI) && endI.After(startTime) {
+			// startTime < endI <= timestamp
+			startTime = endI
+		}
+		if startI.After(timestamp) && startI.Before(endTime) {
```

Comment:
I sort of would like sanity checks that our invariants are actually true.  Life if we run this code on files which have overlapping shards, for instance.  Not a critical wish, just a secondary verification of sanity for the existing shards.

Reply:
We actually already check above that the timestamp we want to add is not contained in an existing shard: 

```
if rpi.ShardGroupByTimestamp(timestamp) != nil {
	return nil
}
```

So we really just walked all the existing shards and validated that there is no shard with `startI <= timestamp < endI`.

And the algo here is robust against existing overlapping shards - we just find the maximum region:
- contained within the original [start, end)
- also containing timestamp
- that doesn't overlap with any existing shards

So if there are existing overlaps we don't fix them, but we don't create any new overlaps.

Reply:
I was thinking we should flag them with a log message.

Reply:
Hmm - I thought that would be a bit noisy for all the customers with existing overlaps, but maybe that is a good idea...

Reply:
Only at Debug level?